### DISCUSSION TITLE
Clarification on constructing maps

### DISCUSSION THREAD

**Discussion by jotaemei** [lfe/lfe#454] [Q&A] [answered]
According to the [tutorial](https://lfe.io/books/tutorial/sequential/propmaps.html), maps can be constructed via 2 methods:

> _As with property lists, maps are a set of key to value associations. You may create an association from "key" to value 42 in one of two ways: using the LFE core form map or entering a map literal:_
![lfe-map-constructions](https://user-images.githubusercontent.com/3826245/174718926-4eb40577-d95b-4e62-bf21-b2ba66f8d20e.png)

The code example on that unit for **tut7.lfe** uses single-quoted atoms for map keys - both in setting them during construction and in pattern matching them in function clauses.

But, this is unreliable, as if a map created via `#M()` is passed to a function, it will not match, for these two constructors vary:

`lfe> (== #M('alpha 14 'beta 17) (map 'alpha 14 'beta 17))`
`false`
`lfe> #M('key 14)`
`#M('key 14)`
`lfe> (map 'key 14)`
`#M(key 14)`

`lfe> (== #M(alpha 14 beta 17) (map 'alpha 14 'beta 17))`  
`true`

The issue, AFAICT, is that, given that `#M()` accepts map literals, the single quote winds up becoming the first character of the key. This is unexpected, however, since the behavior is consistent between the two methods when key names are enclosed in double quotation marks (as exhibited in the tutorial example and in the boolean test below):

`lfe> (== #M("alpha" 14 "beta" 17) (map "alpha" 14 "beta" 17))`
`true`

Edit: In retrospect, my sense is that I should have simply done a pull request for the tutorial and made this clearer there (and perhaps in any documentation), but perhaps this should be left open for discussion.

**Comment by rvirding** [comment, marked as answer]
Right, a couple of points here:

The quote `'exp` is actually expanded at compiletime to the `(quote exp)` form which at runtime guarantees that the expression is never evaluated but just becomes a literal. So writing `(foo a 'b '(+ x y))` will mean that `foo` is called with the arguments the value of the variable `a`, the atom `b` and the list `(+ x y)`. This is just standard Common Lisp which LFE has inherited.

The `(map key val ...)` form is a core form which at runtime creates a map from the key/value arugments. Note that this form evaluates its arguments and then builds the map.

The `#M()` is how a map literal is shown. It can also be used to create a map literal which it done at read time and none of the keys or values are evaluted. It is a literal!

So the `#M()` is not unreliable it is just how maps are written.

This explains why you are getting the bad match: you are doing two different things:
```
lfe> #M('key 14)
#M('key 14)
```
here you have a literal map where the key is the `(quote key)` and the value 14 and as you are writing a literal you are getting the `'key` or `(quote key)`.
```
lfe> (map 'key 14)
#M(key 14)
```
while here you are calling the `map` which at runtime evaluates its arguments and builds the map.

Note that the order of the keys in a map is undefined which is why you will get
```
lfe> #M(c 3 b 2 a 1)
#M(a 1 b 2 c 3)
lfe> (== #M(a 1 b 2 c 3) #M(c 3 b 2 a 1))
true
lfe> (map 'c 99 'x 25 'a 19)
#M(a 19 c 99 x 25)
```
Note that there are no restrictions on the keys and the values in a map, they can be anything. So you can have maps where the keys are tuples or maps:
```
lfe> (map #M(x 19 z 20) 42 (time) (date))
#M(#(18 26 54) #(2022 6 23) #M(x 19 z 20) 42)
```
All these features are properties of the underlying Erlang maps which LFE uses straight off.

Sorry if I went a bit overboard here.
